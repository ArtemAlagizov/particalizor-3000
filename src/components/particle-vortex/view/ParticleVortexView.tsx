import React from "react";
import { Styled } from "../styled";
import { IParticleVortexViewProps } from "./IParticleVortexViewProps";

const ParticleVortexView: React.FC<IParticleVortexViewProps> = ({
  destinationCanvasRef
}) => {
  return (
    <Styled.DestinationCanvas>
      <canvas
        ref={destinationCanvasRef}
        className="particalizor-3000-destination-canvas"
      />
    </Styled.DestinationCanvas>
  );
};

export default ParticleVortexView;
