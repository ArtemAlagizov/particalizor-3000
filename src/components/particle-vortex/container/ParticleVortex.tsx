import React, { useEffect, useRef } from "react";
import ParticleVortexView from "../view/ParticleVortexView";
import { IParticleVortexProps } from "./IParticleVortexProps";
import { getActualParticleVortexValues } from "../../../service/properties-handler/PropertiesHandler";
import {
  clearCanvas,
  drawParticleVortexOnCanvas
} from "../../../service/painter/Painter";
import { createVortexes } from "../../../service/vortex-creator/VortexCreator";
import { createParticles } from "../../../service/particle-creator/ParticleCreator";
import { isNil } from "ramda";

export const ParticleVortex: React.FC<IParticleVortexProps> = ({
  particleNumber,
  particleLifeTime,
  particleTraceWidth,
  imageWidth,
  imageHeight,
  vortexNumber,
  backgroundColor
}) => {
  const destinationCanvasRef = useRef<HTMLCanvasElement>(null);

  const actualValues = getActualParticleVortexValues(
    particleNumber,
    particleLifeTime,
    particleTraceWidth,
    imageWidth,
    imageHeight,
    vortexNumber,
    backgroundColor
  );

  const particles = createParticles(
    actualValues.actualParticleNumber,
    actualValues.actualImageWidth,
    actualValues.actualImageHeight,
    actualValues.actualParticleLifeTime
  );
  const vortexes = createVortexes(
    actualValues.actualVortexNumber,
    actualValues.actualImageWidth,
    actualValues.actualImageHeight
  );

  let animationRequestId: number | null = 0;

  const getAnimationRequestId = () => animationRequestId;
  const setAnimationRequestId = (requestId: number | null) => {
    animationRequestId = requestId;
  };

  useEffect(() => {
    if (!isNil(destinationCanvasRef.current)) {
      drawParticleVortexOnCanvas(
        particles,
        vortexes,
        actualValues,
        destinationCanvasRef.current,
        setAnimationRequestId,
        getAnimationRequestId
      );
    }
  });

  useEffect(() => {
    return () => {
      setAnimationRequestId(null);

      clearCanvas(destinationCanvasRef);
    };
  }, [destinationCanvasRef]);

  return <ParticleVortexView destinationCanvasRef={destinationCanvasRef} />;
};

export default ParticleVortex;
