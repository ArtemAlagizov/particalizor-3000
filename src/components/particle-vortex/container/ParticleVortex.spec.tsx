import React from "react";
import { mount } from "enzyme";
import ParticleVortex from "./ParticleVortex";
import ParticleVortexView from "../view/ParticleVortexView";

describe("Given a ParticleVortex", () => {
  describe("when rendering", () => {
    it("should render ParticleVortexView", () => {
      const component = mount(
        <ParticleVortex imageWidth={100} imageHeight={100} vortexNumber={1} />
      );

      expect(component.find(ParticleVortexView).length).toBe(1);
    });
  });
});
