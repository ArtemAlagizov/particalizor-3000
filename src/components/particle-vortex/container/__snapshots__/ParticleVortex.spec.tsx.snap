// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a ParticleVortex when rendering should match snapshot 1`] = `
<ParticleVortexView
  destinationCanvasRef={
    Object {
      "current": null,
    }
  }
  key="777"
/>
`;
